pkg/webhooks/pod_webhook.go

```diff
@@ -113,16 +113,51 @@ func (p *PodWebhook) Default(ctx context.Context, obj runtime.Object) error {
 		} else {
 			groupUniqueKey = pod.Labels[leaderworkerset.GroupUniqueHashLabelKey]
 		}
-		_, foundEpKey := pod.Annotations[leaderworkerset.ExclusiveKeyAnnotationKey]
-		if foundEpKey && !exclusiveAffinityApplied(*pod) {
-			SetExclusiveAffinities(pod, groupUniqueKey)
+		epKey, foundEpKey := pod.Annotations[leaderworkerset.ExclusiveKeyAnnotationKey]
+		if (foundEpKey) && !exclusiveAffinityApplied(*pod, epKey) {
+			SetExclusiveAffinities(pod, groupUniqueKey, epKey, leaderworkerset.GroupUniqueHashLabelKey)
+		}
+		_, foundSubGroupSize := pod.Annotations[leaderworkerset.SubGroupSizeAnnotationKey]
+		if foundSubGroupSize {
+			// Even if leader does not request TPU resources, this
+			// ensures that it will be scheduled in the same topology
+			// as SubGroup 0
+			pod.Labels[leaderworkerset.SubGroupIndexLabelKey] = "0"
+			subGroupUniqueKey := genGroupUniqueKey(pod.Name, "0")
+			pod.Labels[leaderworkerset.SubGroupUniqueHashLabelKey] = subGroupUniqueKey
+			subEpKey, foundSubEpKey := pod.Annotations[leaderworkerset.SubGroupExclusiveKeyAnnotationKey]
+			if foundSubEpKey && !exclusiveAffinityApplied(*pod, subEpKey) {
+				SetExclusiveAffinities(pod, subGroupUniqueKey, subEpKey, leaderworkerset.SubGroupUniqueHashLabelKey)
+			}
 		}
 	} else {
 		_, workerIndex := statefulsetutils.GetParentNameAndOrdinal(pod.Name)
 		if workerIndex == -1 {
 			return fmt.Errorf("parsing pod ordinal for pod %s", pod.Name)
 		}
 		pod.Labels[leaderworkerset.WorkerIndexLabelKey] = fmt.Sprint(workerIndex)
+		subGroupSize, foundSubGroupSize := pod.Annotations[leaderworkerset.SubGroupSizeAnnotationKey]
+		if foundSubGroupSize {
+			subGroupSizeInt, err := strconv.Atoi(subGroupSize)
+			if err != nil {
+				return err
+			}
+
+			subGroupIndexKey := fmt.Sprint(workerIndex / subGroupSizeInt)
```

Comment:
if the leader is not a worker, and we have 13 pods in total with 3 sub groups. Then I would assume there should be 5 pods in the group 0 right? Then the 5th pod in the group will have worker index as 4, and is actually in subgroup 0. But based on the calculation here, it will be (4/4) = 1, which means it will have sub group index key as 1? am I missing something?

Reply:
nvm, just saw the code below